---
title: "QB25 Week 3 HW Q2"
author: "Joy Mompi"
date: "2025-10-23"
output: html_document
---

```{r setup, include=FALSE}
setwd("~/qb25-answers/week3")
AF_df <- read_delim("~/qb25-answers/week3/AF.tsv", delim = "\t")

ggplot(AF_df, aes(x= Allele_Frequency)) +
  geom_histogram (bins = 11) +
  labs (x = "Allele Frequency") +
  labs (y = "Count") +
  labs (title = "Distribution of Allele Frequencies")

  
#This looks like a normal distribution. Majority of variants cluster around 50% frequency among the samples. I expect the distribution to present this way, since there is approximatley a 50% chance of each  parent passing a particular variant to their offspring.
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
DP_df <- read_delim("~/qb25-answers/week3/DP.tsv", delim = "\t")

ggplot(DP_df, aes(x= Read_Depth)) +
  geom_histogram (bins = 21) +
  xlim(0,20)
  labs (x = "Read Depth") +
  labs (y = "Count") +
  labs (title = "Distribution of Read Depth")
  
#The plot represents the frequency of reads that mapped to each variant in the vcf file. This distribution is skewed right. I'm a bit surprised that the majority of read depths are leaning towards the lower side, since the peak is around 5 reads. I expected the read coverage to be a bit higher overall, but maybe the sequencing technology they used was not as advanced at that time. 
  
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
