---
title: "Week 8 HW"
author: "Joy Mompi"
date: "2025-10-31"
output: html_document
---

```{r setup, include=FALSE}
setwd("~/qb25-answers/week8")
#Step 1.1: Loading data and importing libraries
#First, load the DESeq2, tidyverse, and broom libraries, and use the read_delim() function to read the data and metadata into separate tibbles (the tidyverse version of data frames).
library(tidyverse)
library(DESeq2)
library(broom)

#load counts
wholeblood_counts_df <- read_delim("gtex_whole_blood_counts_downsample.tsv")


#load metadata
gtex_metadata_df <- read_delim("gtex_metadata_downsample.txt")


#load 3rd file
gene_locations_df <- read.delim("~/qb25-answers/week8/gene_locations.txt")


#The gene expression file reports individuals (subjects) in columns and genes in rows. Each entry in the table records the number of reads mapping to that particular gene for the whole-blood sample from that particular subject. Use the function column_to_rownames() to move the gene names, which are stored in the first column to instead be stored as rownames. This is a necessary step due to the format that DESeq2 expects the data, whereby the formal entries in the tibble should be solely numeric.

wholeblood_counts_df[1:5,] 
wholeblood_counts_df <- column_to_rownames(wholeblood_counts_df, var = "GENE_NAME")
wholeblood_counts_df[1:5,]



#The metadata table records the sex, age, and cause of death for each sample. Ages are encoded as integers, ranging from 0 to 4 that represent decade-long age intervals, where 0 denotes ages 20-29, 1 denotes ages 30-39, …, and 5 denotes ages 70-79. By keeping encoding this variable as an integer in our model, rather than a categorical variable, we are modeling linear, monotonic relationships between gene expression level and age.

#Peek at the first few rows and columns of the tibbles, just to understand how they are organized and ensure that they were loaded correctly.

gtex_metadata_df[1:5,]
gtex_metadata_df <- column_to_rownames(gtex_metadata_df, var = "SUBJECT_ID")
gtex_metadata_df[1:5,]


```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
#Step 1.2: Create a DESeq2 object
#It is critical that the columns of the count matrix and the rows of the column data (information about samples) are in the same order. Fortunately, this is already the case for the data we have provided, but it is important to double-check! Write a line or two of code to perform this check.
gtex_metadata_df[1:5,]
row.names(gtex_metadata_df) == colnames(wholeblood_counts_df)

ncol(wholeblood_counts_df)
nrow(gtex_metadata_df)

#Next, load the data into a DESeq2 object using the DESeqDataSetFromMatrix() function. Name this object dds. This function takes three required arguments:
#countData: the gene expression count matrix (or tibble)
#colData: the metadata tibble
#design: the model formula, which uses the linear model notation that we learned from our previous lab on regression, but you can leave off the response variable, since that will always just be the expression of a given gene. For example, if my colData had a variable of interest called CONDITION and another covariate called SMOKING_STATUS, I might want to specify my formula as design = ~ CONDITION + SMOKING_STATUS.

#Replace the brackets below with your code. For your model formula, include the variables SEX, DTHHRDY, and AGE. (The brackets themselves can be removed).

# load DESeq2 object

gtex_dds <- DESeqDataSetFromMatrix(countData = wholeblood_counts_df,
                              colData = gtex_metadata_df,
                              design = ~ SEX + DTHHRDY + AGE)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
#Step 1.3: Normalization and PCA 
#As we did last week, we will want to apply a variance stabilizing transformation (VST) before performing principal component analysis, which has the effect of giving genes with different total expression levels more equal weights. Use the vst() function to apply VST to your DESeq2 object, storing it in a new variable called vsd (for variance-stabilized data).

gtex_vsd <- vst(gtex_dds)

#Now apply the plotPCA() function to your VST-normalized data. Perform this step three separate times, each time providing a different value to the intrgroup = argument to color the points by various variables in your metadata. Save your plots as PNG files and include them in your submission.
png(file = "plotPCA_colorbysex.png")
plotPCA(gtex_vsd, intgroup = "SEX")
dev.off()

png(file = "plotPCA_colorbyage.png")
plotPCA(gtex_vsd, intgroup = "AGE")
dev.off()

png(file = "plotPCA_colorbydeathcondition.png")
plotPCA(gtex_vsd, intgroup = "DTHHRDY")
dev.off()

#What proportion of variance in the gene expression data is explained by each of the first two principal components? Which principal components appear to be associated with which subject-level variables? Interpret these patterns in your own words and record your answers as a comment in your code.

#(Ans 1.3) The first principal component explains 48% of the variance, while the second principal component explains 7% of the variance. It seems like death classification is associated with the first principal component, while age is associated with the second component. I can imagine the whole blood expression values could vary a lot depending on whether they died of natural causes or from a a ondition that required them to be put on a ventilator. Especially those who did not pass from natural causes, likely had specific condition that was marked by particular changes in gene expression. Age also makes sense as a factor, since there are changes to gene expression as one gets older.


```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
```{r}
#Exercise 2: Perform differential expression analysis
#Step 2.1: Perform a “homemade” test for differential expression between the sexes
#You’re curious whether any genes are differentially expressed between sexes (in whole blood). Before you do the DE analysis the “right” way with DESeq2, you’re first going to perform your own naive “homemade” analysis to get an idea of what DESeq2 is doing behind the scenes at a basic level. This will involve regressing expression quantifications from GTEx onto sample sex, and identifying genes with significant differences in expression between sexes. We will use the VST-normalized data for this test, but note that in our later analyses, we will go back to the raw counts!

#As a first step, we can use the code below to extract the VST expression matrix and bind it to the metadata, such that the whole data frame can be used as input to a regression model:

gtex_vsd_df <- assay(gtex_vsd) %>%
  t() %>%
  as_tibble()

gtex_vsd_df <- bind_cols(gtex_metadata_df, gtex_vsd_df)
#For example, we can test for differential expression of the gene WASH7P by running the following code:

m1 <- lm(formula = WASH7P ~ DTHHRDY + AGE + SEX, data = gtex_vsd_df) %>%
  summary() %>%
  tidy()
print(m1)
#Does WASH7P show significant evidence of sex-differential expression (and if so, in which direction)? Explain your answer.

#(Ans 2.1). WASH7P does not show significant evidence of differential expression, becuase the p-value is around .279, which means there is a 27.9% probability that the null hypothesis is true, which is that there are no differences in WASH7P expression as a result of sex.

#Now repeat your analysis for the gene SLC25A47. Does this gene show evidence of sex-differential expression (and if so, in which direction)? Explain your answer.

m2 <- lm(formula = SLC25A47 ~ DTHHRDY + AGE + SEX, data = gtex_vsd_df) %>%
  summary() %>%
  tidy()
print(m2)

#(Ans 2.1)Yes, there is evidence of sex-differential expression of SLC25A47, since the p-value is about .026. There is a 2.6% chance the null hypothesis is true given the results. It seems that there is higher SLC25A47 expression in males compared to female subjects.






```
```{r}
#Step 2.2: Perform differential expression analysis “the right way” with DESeq2
#Now apply the DESeq() function to fit the regression model that you previously specified in your design = argument to each gene in the dataset. This function is a wrapper function that actually performs several different steps. Note that the function should be run on the original un-normalized count data (dds) rather than the data that underwent VST. This is because DESeq2() is designed to model the counts themselves and already includes a step to estimate library sizes (i.e., the total amount of data per sample) and correct for this in the model. Store the results back in the same object, dds.

gtex_dds <- DESeq(gtex_dds)



```

```{r}
#Step 2.3: Extract and interpret the results for sex differential expression
#Results of this differential expression analysis can be extracted using the results() function. By default, this function will focus on whatever variable was specified last in your design = formula, but while controlling for all of the other covariates listed in the model formula. You can override the default behavior and tell results() to focus on any variable of interest that was included in that formula. For example, if my original model formula was design = ~ CONDITION + SMOKING_STATUS, and CONDITION had the levels treatment and control, I could extract differential expression results for CONDITION (controlling for SMOKING_STATUS) and store it in a tibble using the following code.

#Extract the differential expression results for the variable SEX.

gtex_res <- results(gtex_dds, name = "SEX_male_vs_female") %>%
  as_tibble(rownames = "GENE_NAME")
?results
#The padj column of results reports the FDR-adjusted p-value (i.e., “q-value”). The rows with a padj < 0.1 are the genes that are differentially expressed at an FDR of 10%. A lot of rows will have missing padj values. This is because DESeq2 does a filtering step to remove genes that it thinks have a low probability of being DE, and only performs FDR correction for the set of genes it does not filter out. You can ignore genes with missing padj values.

#How many genes exhibit significant differential expression between males and females at a 10% FDR?
gtex_res %>%
  filter(padj < 0.1) %>%
    nrow()
#(Ans) 262 genes have differential expression between males and females at 10% false discovery rate.

#Now load the mappings of genes to chromosomes, which have already downloaded from Dropbox (see top of page). Merge these data with your differential expression results using the left_join() function, setting by = GENE_ID (i.e., the shared column on which you want to join). Order the merged tibble by padj, from smallest to largest.

gene_locations_df <- read.delim("~/qb25-answers/week8/gene_locations.txt")
?left_join()
gtex_res <- gtex_res %>%
  left_join(gene_locations_df, gtex_res, by = "GENE_NAME")

?arrange()
gtex_res <- arrange(gtex_res, padj)


#Examine your top hits. Which chromosomes encode the genes that are most strongly upregulated in males versus females, respectively? Are there more male-upregulated genes or female-upregulated genes near the top of the list? Interpret these results in your own words.

#(Ans) The genes that are most up-regulated in males vs females are located on Chromosome Y, then X, and there are more male-upregulated genes near the top of the list.The Y chromosome is specific to males, so it makes sense that a lot of the sex-mediated variation in expression is due to this. Also, males have only one X chromosome, so they are more prone to X-linked disorders if they have a disease variant in a gene located on X. This could possibly influence the high representation in the reuslts. 


#Examine the results for the two genes (WASH7P and SLC25A47) that you had previously tested with the basic linear regression model in step 2.1. Are the results broadly consistent?
gtex_res %>%
  filter(GENE_NAME == "WASH7P")
#(Ans) Here the p-value I calculated with the linear model was .279 while the adjusted p-value here is .989. It is now even more insignificant after the adjustment for false discovery rate.
gtex_res %>%
  filter(GENE_NAME == "SLC25A47")
#(Ans) The old p-value with the linear model was .026, while the adjusted p-valye is 8.32e-7, so adjusting for false discovery rate enhanced the significance. 

#In your written interpretation of results, add a short reflection (2–3 sentences) on how your analysis illustrates the trade-off between false positives and false negatives. For example, what happens if you use a very stringent FDR threshold (e.g., 1 %) versus a lenient one (e.g., 20 %)? Which type of error (false positive or false negative) would you expect to increase or decrease under each scenario? Briefly comment on how sample size and effect size might influence the power of your analysis to detect truly differentially expressed genes.

#(Ans) If I set the FDR to a higher one like 20%, I'm saying that 20% of the time I reject the null, I expect the null to actually be true. So A large percentage of my significant findings would be "non significant". If the FDR is more stringent, like 1%, then I expect false negatives to increase and false positives to decrease. With a higher FDR, I believe false negatives would decrease and false positives would increase. A lower sample size would give less confidence in true rejections of the mean, so type I errors would increase. Lower effect size could distort the proportion of times one doesn't reject the null hypothesis, when there atcually is siginifant gene expression levels. 

```

```{r}
#Step 2.4: Extract and interpret the results for differential expression by death classification
#Now repeat your analysis from step 2.2 above, but focusing on death classification as your variable of interest. This can be accomplished with the results() function, specifying the argument name = "DTHHRDY_ventilator_case_vs_fast_death_of_natural_causes".

gtex_res2 <- results(gtex_dds, name = "DTHHRDY_ventilator_case_vs_fast_death_of_natural_causes") %>%
  as_tibble(rownames = "GENE_NAME")

gtex_res2 %>%
  filter(padj < 0.1) %>%
    nrow()
#How many genes are differentially expressed according to death classification at a 10% FDR?
#16069 genes
```

```{r}
#Step 2.5: Estimating a false positive rate under the null hypothesis
#To better understand what a false positive looks like in practice, repeat your differential expression analysis but first randomly permute (i.e., shuffle) the sex column from the metadata. Then rerun the DESeq2 analysis using this permuted variable.

gtex_metadata_sh_df <- gtex_metadata_df
gtex_metadata_sh_df$SEX <- sample(gtex_metadata_sh_df$SEX)
  

gtex_dds_sh <- DESeqDataSetFromMatrix(countData = wholeblood_counts_df,
                              colData = gtex_metadata_sh_df,
                              design = ~ SEX + DTHHRDY + AGE)

gtex_dds_sh <- DESeq(gtex_dds_sh)

gtex_res_sh <- results(gtex_dds_sh, name = "SEX_male_vs_female") %>%
  as_tibble(rownames = "GENE_NAME")


#How many genes appear “significant” in this permuted analysis at a 10 % FDR? These are, by definition, false positives, because the labels no longer correspond to real biological sex. Compare this number to the count from your real (non-permuted) analysis. What does this suggest about how well the FDR threshold controls the expected rate of false discoveries in large-scale RNA-seq experiments?
gtex_res_sh %>%
  filter(padj < 0.1) %>%
    nrow()

#(Ans) Only 3 genes appear significant here, compared to 262 in the non-permuted version. I can assume that setting a fdr at 10% does an accurate job of detecting the proportion of false positive rates in our results.  
```
```{r}
#Exercise 3: Visualization
#Use ggplot2 to create a “Volcano” plot depicting your differential expression results from Exercise 2.3. A volcano plot is a scatter plot, where the x-axis shows the log2FoldChange and the y-axis shows the -log10(padj).
#Highlight the genes that are significant at a 10% FDR AND for which the absolute value of the log2FoldChange is greater than 1 in a separate color.
threshold <- .1
threshold2 <- 1
#got the above code from copilot, where it assigns values to a threshold and input them into my plot 

volcano_plot <- ggplot(data = gtex_res, aes(x = log2FoldChange, y = -log10(padj), color = gtex_res$padj < threshold & gtex_res$log2FoldChange > threshold2)) + 
  geom_point() +
  xlim(-10, 10)

  
#Highlight the genes that are significant at a 10% FDR AND for which the absolute value of the log2FoldChange is greater than 1 in a separate color.

#Output this plot to a .png that you will upload with your assignment.
ggsave("volcano_plot.png", plot = volcano_plot, path = "~/qb25-answers/week8")
```

